mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala

```diff
@@ -0,0 +1,225 @@
+   * Loss function which GBT tries to minimize. (case-insensitive)
+   * Supported: "LogLoss"
+   * (default = LogLoss)
+   * @group param
+   */
+  val loss: Param[String] = new Param[String](this, "loss", "Loss function which GBT tries to" +
+    " minimize (case-insensitive). Supported options: LogLoss")
+
+  setDefault(loss -> "logloss")
+
+  /** @group setParam */
+  def setLoss(value: String): this.type = {
+    val lossStr = value.toLowerCase
+    require(GBTClassifier.supportedLosses.contains(lossStr), "GBTClassifier was given bad loss:" +
+      s" $value. Supported options: ${GBTClassifier.supportedLosses.mkString(", ")}")
+    set(loss, lossStr)
+    this
+  }
+
+  /** @group getParam */
+  def getLoss: String = getOrDefault(loss)
+
+  /** (private[ml]) Convert new loss to old loss. */
+  override private[ml] def getOldLoss: OldLoss = {
+    getLoss match {
+      case "logloss" => OldLogLoss
+      case _ =>
+        // Should never happen because of check in setter method.
+        throw new RuntimeException(s"GBTClassifier was given bad loss: $getLoss")
+    }
+  }
+
+  override protected def train(
+      dataset: DataFrame,
+      paramMap: ParamMap): GBTClassificationModel = {
+    val categoricalFeatures: Map[Int, Int] =
+      MetadataUtils.getCategoricalFeatures(dataset.schema(paramMap(featuresCol)))
+    val numClasses: Int = MetadataUtils.getNumClasses(dataset.schema(paramMap(labelCol))) match {
+      case Some(n: Int) => n
+      case None => throw new IllegalArgumentException("GBTClassifier was given input" +
+        s" with invalid label column, without the number of classes specified.")
+      // TODO: Automatically index labels.
+    }
+    require(numClasses == 2,
+      s"GBTClassifier only supports binary classification but was given numClasses = $numClasses")
+    val oldDataset: RDD[LabeledPoint] = extractLabeledPoints(dataset, paramMap)
+    val boostingStrategy = super.getOldBoostingStrategy(categoricalFeatures, OldAlgo.Classification)
+    val oldGBT = new OldGBT(boostingStrategy)
+    val oldModel = oldGBT.run(oldDataset)
+    GBTClassificationModel.fromOld(oldModel, this, paramMap, categoricalFeatures)
+  }
+}
+
+object GBTClassifier {
+  // The losses below should be lowercase.
+  /** Accessor for supported loss settings */
+  final val supportedLosses: Array[String] = Array("logloss").map(_.toLowerCase)
+}
+
+/**
+ * :: AlphaComponent ::
+ *
+ * [[http://en.wikipedia.org/wiki/Gradient_boosting Gradient-Boosted Trees (GBTs)]]
+ * model for classification.
+ * It supports binary labels, as well as both continuous and categorical features.
+ * Note: Multiclass labels are not currently supported.
+ * @param trees  Decision trees in the ensemble.
+ * @param treeWeights  Weights for the decision trees in the ensemble.
+ */
+@AlphaComponent
+final class GBTClassificationModel(
+    override val parent: GBTClassifier,
+    override val fittingParamMap: ParamMap,
+    val trees: Array[DecisionTreeRegressionModel],
```

Comment:
Sort of.  Gradient boosting's "gradient" part comes from phrasing the problem as regression.

scikit-learn trains 1 regression tree for each class, essentially doing a 1 vs all approach to reduce multiclass to binary.  Those predictions can then be fed into the boosting algorithm via a multinomial logistic loss function.

Just as in reducing multiclass to binary, we can take other approaches.  E.g., if you use an error-correcting code-based approach, then it can make sense to train 1 tree per iteration, where each tree tries to separate some subset of classes out from the other classes.

I have seen (a) scikit-learn's approach of 1-vs-all for gradient boosting and (b) the error-correcting approach for other boosting algorithms, but I have not seen someone write out the combination: gradient boosting + error-correcting.  It may be out there, and I'm sure we could work it out, with a bit of effort.

So, this abstraction will work if we take the error-correcting approach.  It could work for 1-vs-all if we keep a matching array indicating which class each tree is testing for, but then the ordering of the trees would be obscured.  For 1-vs-all, it might be best to store ```Array[Array[DecisionTreeRegressionModel]]```